src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,287 @@
+        for (auto& log_entry: _streamLog){
+            readFutures.push_back(_client.read(log_entry.name, 0, log_entry.offset));
+        }
+        return seastar::when_all_succeed(readFutures.begin(), readFutures.end());
+    })
+    .then([this] (std::vector<std::tuple<Status, Payload> >&& responses){
+        std::vector<Payload> newPayloads;
+        for (auto& response: responses){
+            auto& [status, resp] = response;
+            if (!status.is2xxOK()){
+                throw std::runtime_error("unable to read a plog");
+            }
+            newPayloads.push_back(std::move(resp));
+        }
+        return seastar::make_ready_future<std::vector<Payload> >(std::move(newPayloads));
+    });
+}
+
+
+seastar::future<> 
+LogStream::write(Payload payload){
+    return _write(std::move(payload), 1);
+}
+
+seastar::future<> 
+LogStream::_write(Payload payload, uint32_t writeToWAL){
+    if (!_logStreamCreate){
+        K2INFO("Log stream does not create");
+        throw std::runtime_error("Log stream does not create");
+    }
+
+    String sealed_plogId;
+    String new_plogId;
+    uint32_t sealed_offest = 0;
+    
+    // determine whether the current plog will excceed the size limit after the incoming writing operation
+    if (_plogInfo[writeToWAL].second + payload.getSize() >= PLOG_MAX_SIZE){
+        sealed_plogId = _plogInfo[writeToWAL].first;
+        sealed_offest = _plogInfo[writeToWAL].second;
+        // change to the next available plog id
+        if (writeToWAL == 0){
+            new_plogId = _metadataPlogPool[_metadataPlogPool.size()-1];
+            _metadataPlogPool.pop_back();
+        }
+        else{
+            new_plogId = _walPlogPool[_walPlogPool.size()-1];
+            _walPlogPool.pop_back();
+        }
+        _plogInfo[writeToWAL].first = new_plogId;
+        _plogInfo[writeToWAL].second = 0;
+    }
+
+    uint32_t current_offset = _plogInfo[writeToWAL].second;
+    _plogInfo[writeToWAL].second += payload.getSize();
+
+    // write the incoming payload to current plog
+    return _client.append(_plogInfo[writeToWAL].first, current_offset,  std::move(payload))
+    .then([&, writeToWAL, sealed_plogId, new_plogId, sealed_offest] (auto&& response){
+        auto& [status, resp] = response;
+        if (!status.is2xxOK()) {
+            K2INFO("unable to append the plog");
+            K2INFO(status);
+            throw std::runtime_error("unable to append the plog");
+        }
+        // If sealed_offest == 0, then it means after the incoming writing operation, the current plog is not excceed the size limit
+        if (sealed_offest == 0){
+            return seastar::make_ready_future<>();
+        }
+        // It the previous plog excceed the size limit, we will seal the previous plog and write the metadata to corrsponding place (Metadata Log Stream or CPO)
+        return _client.seal(sealed_plogId, sealed_offest)

```

Comment:
I think it is better to seal first before doing the append.

Reply:
If we do the seal first, we need to wait for this seal to be done before we execute the next writing. I think this seems to be a lock, which we would like to avoid. And do the append first will not affect the correctness of this scenario, and can increase the throughput.

Reply:
Maybe, I think you need to at least write to the metadata log first then, maybe Ivan or Jerry can help clarify.

Reply:
If there is an ongoing seal request, I will first send the upcoming write request to the plog servers. However, if this write operation finished, it will not return success to the client until the ongoing seal request finished.